pennylane/templates/subroutines/qrom.py

```diff
@@ -0,0 +1,295 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This submodule contains the template for QROM.
+"""
+
+import math
+
+import numpy as np
+
+import pennylane as qml
+from pennylane.operation import Operation
+
+
+def _multi_swap(wires1, wires2):
+    """Apply a series of SWAP gates between two sets of wires."""
+    for wire1, wire2 in zip(wires1, wires2):
+        qml.SWAP(wires=[wire1, wire2])
+
+
+class QROM(Operation):
+    r"""Applies the QROM operator.
+
+    This operator encodes bitstrings associated with indexes:
+
+    .. math::
+        \text{QROM}|i\rangle|0\rangle = |i\rangle |b_i\rangle,
+
+    where :math:`b_i` is the bitstring associated with index :math:`i`.
+
+    Args:
+        bitstrings (list[str]): the bitstrings to be encoded
+        control_wires (Sequence[int]): the wires where the indexes are specified
+        target_wires (Sequence[int]): the wires where the bitstring is loaded
+        work_wires (Sequence[int]): the auxiliary wires used for the computation
+        clean (bool): if True, the work wires are not altered by operator, default is ``True``
+
+    **Example**
+
+    In this example, the QROM operator is applied to encode the third bitstring, associated with index 2, in the target wires.
+
+    .. code-block::
+
+        # a list of bitstrings is defined
+        bitstrings = ["010", "111", "110", "000"]
+
+        dev = qml.device("default.qubit", shots = 1)
+
+        @qml.qnode(dev)
+        def circuit():
+
+            # the third index is encoded in the control wires [0, 1]
+            qml.BasisEmbedding(2, wires = [0,1])
+
+            qml.QROM(bitstrings = bitstrings,
+                    control_wires = [0,1],
+                    target_wires = [2,3,4],
+                    work_wires = [5,6,7])
+
+          return qml.sample(wires = [2,3,4])
+
+    .. code-block:: pycon
+
+        >>> print(circuit())
+        [1 1 0]
+
+
+    .. details::
+        :title: Usage Details
+
+        This template takes as input three different sets of wires. The first one is ``control_wires`` which is used
+        to encode the desired index. Therefore, if we have :math:`m` bitstrings, we need
+        at least :math:`\lceil \log_2(m)\rceil` control wires.
+
+        The second set of wires is ``target_wires`` which stores the bitstrings.
+        For instance, if the bitstring is "0110", we will need four target wires. Internally, the bitstrings are
+        encoded using the :class:`~.BasisEmbedding` template.
+
+
+        The ``work_wires`` are the auxiliary qubits used by the template to reduce the number of gates required.
+        Let :math:`k` be the number of work wires. If :math:`k = 0`, the template is equivalent to executing :class:`~.Select`.
+        Following the idea in `arXiv:1812.00954 <https://arxiv.org/abs/1812.00954>`__, auxiliary qubits can be used to
```

Comment:
Just double check what convention we use when adding links to ArXiv, in documentation.

Reply:
For example, in the QSVT documentation we apply the link to the name `quantum singular value transformation` or we add link with the name of the paper `Grand Unification of Quantum Algorithms ...`

Reply:
Just updated to follow the same format we decided on in AmplitudeAmplification